rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,278 @@
+access requests.
+
+Users won’t be asked to think about roles in advance, but will be able to focus
+on finding and requesting access to individual resources.
+
+Both “role-based” and “search-based” access requests are useful for two
+different scenarios:
+
+- Sometimes it is helpful to request elevated roles in the system, like
+  `db-admin`, to perform some system upgrades, or system wide-troubleshooting.
+  In this case, role-based access requests work great.
+
+- In other cases, it’s helpful to request access to a subset of resources. In
+  this case, search.
+
+### Example Flow
+
+Alice is a member of the group “splunk”.
+
+She gets an alert from a node “db-1”.
+
+Alice needs to locate the node and request access to it.
+
+Alice goes to the “Request access” screen and selects: “select access to resources”
+option that is available alongside the “select access to roles” option.
+
+Alice uses standard (fuzzy search) or advanced search (labels) to enter the
+resource name.
+
+She finds a single resource or a group of resources by that name and clicks
+"request access".
+
+Teleport identifies that access to the node “db-1” is managed by team
+“db-admins” based on Alice's role which allowed her to search for this node.
+
+An access request is created and sent to the "db-admins" team members via slack,
+email or other configured channel.
+
+There are two thresholds set in access requests for this group of resources.
+When both Ivan and Mary approve the request, Alice is granted access for one
+hour.
+
+### Query Language and Filter UI
+
+The query language and filter UI is defined by the
+[Pagination and Search RFD](./0055-webui-ss-paginate-filter.md).
+
+### CLI Mock-Up
+
+```bash
+$ tsh request search --kind db --labels 'env=prod'
+Found 2 items:
+
+name kind     id
+db-1 database 388aff7f-459f-4a43-804a-3729854976ab
+db-2 database 3be2fdad-7c79-4cfa-924e-ec1ea7225320
+
+Create access request by:
+> tsh request create --id 388aff7f-459f-4a43-804a-3729854976ab --id 3be2fdad-7c79-4cfa-924e-ec1ea7225320
+```
+
+Users can search by kind, labels, and keywords. The `tsh request search` command
+will output a `tsh request create` command which can be used to request access.
+
+The CLI UX may be improved by allowing users to perform multiple searches,
+"stage" which resources they want to request access to and store this state
+locally, and finally execute the full request without needing each UUID on the
+command line. But this may be too complex and confusing. With this MVP version
+users can copy-paste UUIDs from multiple searches, the meaning is clear, and it
+is fully customizable and scriptable.
+
+### On-Demand SSH
+
+Many times users would not want to search and request access in two steps.
+Teleport will have a new alias flag modifier that will try accessing a node, in
+case if access denied, it will request access for a node all in one step:
+
+```bash
+# -P "Please" activate try and request mode
+$ tsh ssh -P root@db-node
```

Comment:
Agreed, I just copied this from the google doc for now. This behaviour is interesting enough that it probably warrants a long flag, I would propose `--request`

Reply:
Having to ssh with a specific flag sort of defeats the purpose of this functionality though.  Additionally, if this workflow will support openssh can an informational message be generated for the user so they know their access is denied but an access request is generated and pending approval (though not sure whether openssh will support that).

Reply:
@jdconti see the section with aliases that will allow to have `tsh ssh` to work like `tsh ssh -P` for your org

Reply:
^^ +1; if we can enable this by default that would be great. if we cant, can you have an env variable that enables it so we can (try) to set it globally for our users? AS noted above, its really ideal for us that our users dont have to remember to do anything different (different flag, different command, different workflow) to access this new type of hosts if they are using tsh ssh.

Reply:
@jdconti @daviesalex @klizhentas I'm wary of changing the default behaviour globally and potentially causing people to accidentally create access requests, e.g. if they merely mistyped a hostname or something.

Would it be acceptable to have a prompt or a suggested command if the user cannot currently access the host but would be able to request access to it? Something like the following:
```bash
$ tsh ssh ubuntu@host1
ERROR: access denied to ubuntu connecting to host1 on cluster cluster-one

Request access (Y/n)? Y
Enter request reason: need to fix x
Creating request...
Request ID: c723ff91-536b-4125-b2bd-437aafa5ea15
Username:   ubuntu
Roles:      splunk
Resources:  388aff7f-459f-4a43-804a-3729854976ab (host1)
Reason:     "need to fix x"
Reviewers:  [none] (suggested)
Status:     PENDING

hint: use 'tsh login --request-id=<request-id>' to login with an approved request

Waiting for request approval...

Approval received, getting updated certificates...

ubuntu@host1:~$
```

Reply:
@daviesalex with aliases (see https://github.com/gravitational/teleport/pull/11447/files) you will be able to set this globally for your team

Reply:
@nklaassen the flow (Y/N) will defeat the purpose of `-R` that implies that is what users want (create a request if they can't access the node), so I think we don't need it

Reply:
@nklaassen thanks for your detail. Your workflow is I think perfect. 

Its not ideal that we would have to manage a file in every users environment, but I totally understand the desire to not confuse users [though would it be right that if users dont enable the new feature by not having any users in this new category of 'need to ask for permission for this host', the behaviour is not changing?]. In any case, I think we can deal with this. 

cc @jdconti for thoughts?

Reply:
One wrinkle I can see here is how to handle the availability of multiple roles that may be requested. If a user's role has multiple `search_as_roles` values which might apply to a host or the user has multiple roles each with `search_as_roles` values which might apply to the host, which are applied when the access request is granted?

Reply:
> Its not ideal that we would have to manage a file in every users environment

I agree, it would be helpful to support other methods of enabling automatic access request creation (e.g. environment, existing role, and perhaps user traits/metadata/claims).  While dotfiles are great for opinionated users managing them at scale can be painful; supporting a global `/etc/tshconfig` would ease that pain slightly (as I mentioned in #11447)

Reply:
> One wrinkle I can see here is how to handle the availability of multiple roles that may be requested. If a user's role has multiple search_as_roles values which might apply to a host or the user has multiple roles each with search_as_roles values which might apply to the host, which are applied when the access request is granted?

If the user has multiple `search_as_roles` which could grant access to the requested resource, all such roles would be requested by default (with access limited to the specific resource requested). If for some reason a specific role is needed (which the user would know about in this case) the web UI and the `tsh request create` command will provide the option to request a specific role with a search-based access request.

Reply:
@nklaassen to follow the principle of least privilege, we should request access to the minimal role set that grants the privileges, for example  if Alice is executing `ssh alice@host`, and there are two roles: `a` -> grants access to `host` as `alice` and `b` that grants access as `root, alice`, only `a` should be requested.

Reply:
@klizhentas it should be pretty straightforward to do this for `tsh ssh --request user@node` requests where we know the login, we can just choose a role with the minimal set of logins. When using the actual search flow we could try to do something similar by making the user put in the set of logins they need, but I think the simplest solution for that case is still to request all available search_as_roles by default, the cert will be limited to only the requested resources.